<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [Twisted-Python] Understanding deferred and error handling
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:twisted-python%40twistedmatrix.com?Subject=%5BTwisted-Python%5D%20Understanding%20deferred%20and%20error%20handling&In-Reply-To=4DAE60E9.80903%40silveregg.co.jp">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="023896.html">
   <LINK REL="Next"  HREF="023898.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[Twisted-Python] Understanding deferred and error handling</H1>
    <B>Jason Rennie</B> 
    <A HREF="mailto:twisted-python%40twistedmatrix.com?Subject=%5BTwisted-Python%5D%20Understanding%20deferred%20and%20error%20handling&In-Reply-To=4DAE60E9.80903%40silveregg.co.jp"
       TITLE="[Twisted-Python] Understanding deferred and error handling">jrennie at gmail.com
       </A><BR>
    <I>Wed Apr 20 06:15:08 EDT 2011</I>
    <P><UL>
        <LI>Previous message: <A HREF="023896.html">[Twisted-Python] Understanding deferred and error handling
</A></li>
        <LI>Next message: <A HREF="023898.html">[Twisted-Python] Understanding deferred and error handling
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#23897">[ date ]</a>
              <a href="thread.html#23897">[ thread ]</a>
              <a href="subject.html#23897">[ subject ]</a>
              <a href="author.html#23897">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>On Wed, Apr 20, 2011 at 12:28 AM, David &lt;<A HREF="http://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">david at silveregg.co.jp</A>&gt; wrote:

&gt;<i>     def _stop(arg):
</I>&gt;<i>         reactor.stop()
</I>&gt;<i>     d.addBoth(_stop)
</I>&gt;<i>
</I>
Try using addCallbacks instead of addBoth.  Then, you can logically separate
code to handle/print errors from &quot;normal&quot; code.

This will simply print no error in the log:
&gt;<i>
</I>
That's because you discarded the twisted.python.failure.Failure object by
using _stop for your errback and not doing anything with the arg.  Similar
to try/except with a &quot;pass&quot; in the &quot;except&quot; clause.

    def log_error(failure):

        log.err(failure.printTraceback())
&gt;<i>         return failure
</I>&gt;<i>     d.addErrback(log_error)
</I>&gt;<i>     d.addBoth(_stop)
</I>&gt;<i>
</I>
Use d.addCallbacks(_stop, log_error) instead of addBoth/addErrback.  Also,
you probably want &quot;log.err(failure.getTraceback())&quot; instead of
&quot;log.err(failure.printTraceback())&quot;.  printTraceback does not return a
meaningful value IIUC.

I don't get any traceback either.
&gt;<i>
</I>
Your traceback went to stdout whereas you were probably watching stderr or a
log file.

Jason

-- 
Jason Rennie
Research Scientist, ITA Software
617-714-2645
<A HREF="http://www.itasoftware.com/">http://www.itasoftware.com/</A>
-------------- next part --------------
An HTML attachment was scrubbed...
URL: <A HREF="http://twistedmatrix.com/pipermail/twisted-python/attachments/20110420/6708b033/attachment.htm">http://twistedmatrix.com/pipermail/twisted-python/attachments/20110420/6708b033/attachment.htm</A> 
</PRE>














<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="023896.html">[Twisted-Python] Understanding deferred and error handling
</A></li>
	<LI>Next message: <A HREF="023898.html">[Twisted-Python] Understanding deferred and error handling
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#23897">[ date ]</a>
              <a href="thread.html#23897">[ thread ]</a>
              <a href="subject.html#23897">[ subject ]</a>
              <a href="author.html#23897">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">More information about the Twisted-Python
mailing list</a><br>
</body></html>
